haystack/preview/dataclasses/document.py

```diff
@@ -0,0 +1,92 @@
+from typing import List, Any, Dict, Literal, Optional
+
+import json
+import hashlib
+import logging
+from pathlib import Path
+from dataclasses import asdict, dataclass, field
+
+logger = logging.getLogger(__name__)
+
+try:
+    from numpy import ndarray
+except ImportError as exc2:
+    logger.debug("numpy can't be imported. You won't be able to use embeddings.")
+    ndarray = None  # type: ignore
+
+try:
+    from pandas import DataFrame
+except ImportError as exc3:
+    logger.debug("pandas can't be imported. You won't be able to use table related features.")
+    DataFrame = None  # type: ignore
+
+
+ContentTypes = Literal["text", "table", "image"]
+
+
+CONTENT_TYPES: Dict[ContentTypes, type] = {"text": str, "table": DataFrame, "image": Path}
+
+
+@dataclass(frozen=True)
+class Document:
+    """
+    Base data class containing some data to be queried.
+    Can contain text snippets, tables, file paths to images.
+    Documents can be sorted by score, serialized to/from dictionary and JSON, and are immutable.
+
+    Note that `id_hash_keys` are referring to keys in the metadata. `content` is always included in the id hash.
+    """
+
+    id: str = field(default_factory=str)
+    content: Any = field(default_factory=lambda: None)
+    content_type: ContentTypes = "text"
+    metadata: Dict[str, Any] = field(default_factory=dict, hash=False)
+    id_hash_keys: List[str] = field(default_factory=lambda: [], hash=False)
+    score: Optional[float] = field(default=None, compare=True)
+    embedding: Optional[ndarray] = field(default=None, repr=False)
+
+    def __str__(self):
+        return f"{self.__class__.__name__}('{self.content}')"
+
+    def __post_init__(self):
+        """
+        Generate the ID based on the init parameters and make sure that content_type
+        matches the actual type of content.
+        """
+        # Validate content_type
+        if not isinstance(self.content, CONTENT_TYPES[self.content_type]):
+            raise ValueError(
+                f"The type of content ({type(self.content)}) does not match the "
+                f"content type: '{self.content_type}' expects '{CONTENT_TYPES[self.content_type]}'."
+            )
+        # Check if id_hash_keys are all present in the meta
+        for key in self.id_hash_keys:
+            if key not in self.metadata:
+                raise ValueError(
+                    f"'{key}' must be present in the metadata of the Document if you want to use it to generate the ID."
+                )
+        # Generate the ID
+        content_to_hash = ":".join(
+            [
+                self.__class__.__name__,
+                str(self.content),
+                *[str(self.metadata.get(key, "")) for key in self.id_hash_keys],
+            ]
+        )
+        hashed_content = hashlib.sha256(str(content_to_hash).encode("utf-8")).hexdigest()
+        object.__setattr__(self, "id", hashed_content)
```

Comment:
why not `setattr(self, "id", hashed_content)`?

Reply:
Great question :grin: That's because `Document` is a frozen dataclass, i.e. "immutable".

Feel free to question this experiment, btw. My idea is that `Documents` should be immutable because of the `id_hash_keys` attribute: so if the ID is generated by hashing the values of `content` and `meta`, any change to `content` or `meta` should also affect the ID. Given that re-computing the ID in these situation would mean making a property out of it, and therefore slowing the class on a very frequent code path, I opted to make the class immutable and force users to create a new document every time they want to change the content or meta.

I don't know yet how impactful this difference is going to be. Happy to revert (now or later)
 if we realize that this technique introduces more pain than it removes. The good news is that making this class mutable again is really straightforward.